OpenRA.Mods.Common/Traits/World/AutoSave.cs

```diff
@@ -0,0 +1,99 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[TraitLocation(SystemActors.World)]
+	[Desc("Add this trait to the world actor to enable auto-save.")]
+	public class AutoSaveInfo : TraitInfo
+	{
+		public override object Create(ActorInitializer init) { return new AutoSave(init.Self, this); }
+	}
+
+	public class AutoSave : ITick
+	{
+		const string AutoSavePattern = "autosave-";
+		const string SaveFileExtension = ".orasav";
+		int ticksUntilAutoSave;
+		int lastSaveInverval;
+		readonly bool isDisabled;
+
+		public AutoSave(Actor self, AutoSaveInfo info)
+		{
+			ticksUntilAutoSave = GetTicksBetweenAutosaves(self);
+			lastSaveInverval = Game.Settings.SinglePlayerSettings.AutoSaveInterval;
+
+			isDisabled = self.World.LobbyInfo.GlobalSettings.Dedicated || self.World.LobbyInfo.NonBotClients.Count() > 1;
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (isDisabled || self.World.IsReplay || self.World.IsLoadingGameSave)
+				return;
+
+			var autoSaveValue = Game.Settings.SinglePlayerSettings.AutoSaveInterval;
+
+			if (autoSaveValue == 0)
+				return;
+
+			var autoSaveFileLimit = Game.Settings.SinglePlayerSettings.AutoSaveFileNumber;
+
+			autoSaveFileLimit = autoSaveFileLimit < 3 ? 3 : autoSaveFileLimit;
+
+			if (lastSaveInverval != autoSaveValue)
+			{
+				lastSaveInverval = autoSaveValue;
+				ticksUntilAutoSave = GetTicksBetweenAutosaves(self);
+			}
+
+			if (--ticksUntilAutoSave > 0)
+				return;
+
+			var oldAutoSaveFiles = GetAutoSaveFiles()
+				.OrderByDescending(f => f.CreationTime)
+				.Skip(autoSaveFileLimit - 1);
+
+			foreach (var oldAutoSaveFile in oldAutoSaveFiles)
+				oldAutoSaveFile.Delete();
```

Comment:
I'm afraid that this might start to hurt performance if interval is low and file count high. 

Potentionally you'll be querying a huge file list and then each time deleting only one file.

Perhaps it's ok.

Might be best to force a minimum interval of at least 30 or even 60 seconds.

Reply:
do you want this to happen on a separate thread? If user selects a sertain number of saves he does expect that number to be maintained